### PR TITLE
Add context for comments in HTML html_tokenizer

### PR DESCRIPTION
**Pull request by feliperenan**
This PR adds a context `[:comment_start]` and `[:comment_end]` to tokens that are HTML comments. Both will be added in case multiple comments start in the same line:

```html
<!-- <%= 1 %> --><!-- <%= 2 %> -->
```

In this case we would have `[:comment_start]`, `[:comment_end, :comment_start]`, `[:comment_end]`.

### CODE REVIEW

File: lib/phoenix_live_view/html_tokenizer.ex [hunk 1]
```diff
@@ -94,11 +94,12 @@ defmodule Phoenix.LiveView.HTMLTokenizer do
   end
 
   defp handle_text("<!--" <> rest, line, column, buffer, acc, state) do
+    state = Map.put(state, :context, :comment_start)
```
Comment (feliperenan) [inline lib/phoenix_live_view/html_tokenizer.ex:97]: @josevalim I'm not 100% sure if that is the correct approach.  I will wait for your feedback so I can do the same for the `script` tag 🙌🏻  and finish this PR.

Comment (feliperenan) [reply lib/phoenix_live_view/html_tokenizer.ex:97]: I added in the state to get it working faster but I guess we shouldn't do it otherwise we might have the wrong state in case we don't clean up. Maybe a better way would be changing the function signature to something like:

```elixir
  defp handle_text("<!--" <> rest, meta, buffer, acc, state) do
    meta = %{line: line, column: column, context: :comment}
    handle_comment(rest, meta, ["<!--" | buffer], acc, state)
  end
```

But that would require more changes since we would need to change all functions.

Comment (josevalim) [reply lib/phoenix_live_view/html_tokenizer.ex:97]: I think you can keep it in the state, as it is rarely updated. However, you need to keep all starting and ending tokens, because in the same text you can have HTML comment starting and finishing. And then you prune it down at the end. For example, when the comment starts, you would do:

```elixir
put_in state.context, &[:comment_start | &1]
```

And when it finishes:

```
put_in state.context, &[:comment_end | &1]
```

And then at the end, you want to pop it from the state and remove closed contexts:

```elixir
defp reverse_context([:comment_end, :comment_start | rest], acc), do: reverse_context(rest, acc)
defp reverse_context([:script_end, :script_start | rest], acc), do: reverse_context(rest, acc)
defp reverse_context([head | rest], acc), do: reverse_context(rest, [head | acc])
defp reverse_context([], acc), do: acc
```

And then you add the context to meta if it is not an empty list. The complicated tests are stuff like this:

1. Comment starts and end in the same text

```html
Some <!-- comment --!> text
```

2. Comment ends and script starts in the same text

```html
<!-- <%= "Hello" %> --><script><%= "World" %></script>
```

File: test/phoenix_live_view/html_tokenizer_test.exs [hunk 2]
```diff
@@ -64,11 +65,75 @@ defmodule Phoenix.LiveView.HTMLTokenizerTest do
...
+               {:tag_open, "p", [], %{column: 1, line: 1}},
+               {:text, "\n<!--\n<div>\n",
+                %{column_end: 1, line_end: 4, context: [:comment_start]}},
+               {:text, "</div>\n-->\n", %{column_end: 1, context: [:comment_end], line_end: 3}},
+               {:tag_close, "p", %{column: 1, line: 3}},
+               {:text, "\n", %{column_end: 1, line_end: 4}}
+             ]
+    end
+
+    test "handle comment and with script" do
```
Comment (feliperenan) [inline test/phoenix_live_view/html_tokenizer_test.exs:109]: here I'm trying to reproduce the scenario where parse comment + eex + script as you pointed out.  Here is how it is tokenized:

```elixir
iex(> EEx.Tokenizer.tokenize(contents, 0, 0, %{indentation: 0, trim: false})
{:ok,
 [
   {:text, 0, 0, '<!-- '},
   {:expr, 0, 5, '=', ' "Hello" '},
   {:text, 0, 19, ' --><script>'},
   {:expr, 0, 31, '=', ' "World" '},
   {:text, 0, 45, '</script>'},
   {:eof, 0, 54}
 ]}
```

Because we append eex expressions into comments and script there in the Formatter, that is exactly how it would be processed here in the `HTMLTokenizer`.

Also, notice that different from comments, script has tag open and close. So I believe we don't need any metadata for them.

File: test/phoenix_live_view/html_tokenizer_test.exs [hunk 3]
```diff
@@ -64,11 +65,75 @@ defmodule Phoenix.LiveView.HTMLTokenizerTest do
 
       assert [
                {:tag_open, "p", [], %{line: 1, column: 1}},
-               {:text, "\n<!--\n<div>\n-->\n", %{line_end: 5, column_end: 1}},
+               {:text, "\n<!--\n<div>\n-->\n",
+                %{line_end: 5, column_end: 1, context: [:comment_start, :comment_end]}},
```
Comment (feliperenan) [inline test/phoenix_live_view/html_tokenizer_test.exs:69]: @josevalim my code is not yet 100% I just wanted to update the test cases to make sure what we are expecting here. 

If I get it right, we want to add `comment_start` and `comment_end` when it is inline comments. Does it make sense? 🤔

File: test/phoenix_live_view/html_tokenizer_test.exs [hunk 4]
```diff
@@ -64,11 +65,75 @@ defmodule Phoenix.LiveView.HTMLTokenizerTest do
...
+      </p>
+      """
+
+      {tokens, :text} = HTMLTokenizer.tokenize(second_part, "nofile", 0, [], tokens, cont)
+
+      assert Enum.reverse(tokens) == [
+               {:tag_open, "p", [], %{column: 1, line: 1}},
+               {:text, "\n<!--\n<div>\n",
+                %{column_end: 1, line_end: 4, context: [:comment_start]}},
+               {:text, "</div>\n-->\n", %{column_end: 1, context: [:comment_end], line_end: 3}},
```
Comment (feliperenan) [inline test/phoenix_live_view/html_tokenizer_test.exs:103]: And that is how it would like for cases that the comment ends in the next token.

File: test/phoenix_live_view/html_tokenizer_test.exs (hunk 2 above)
Comment (josevalim) [reply test/phoenix_live_view/html_tokenizer_test.exs:109]: The specific details are not relevant IMO. The important is that you can achieve what you need in the formatter. :)

File: test/phoenix_live_view/html_tokenizer_test.exs (hunk 3 above)
Comment (josevalim) [reply test/phoenix_live_view/html_tokenizer_test.exs:69]: In this case they should cancel out IMO and there is no context. Unless you need to know that there was a comment at some point?

Comment (feliperenan) [reply test/phoenix_live_view/html_tokenizer_test.exs:69]: This would be nice to have but as just `:comment`. But I wouldn't worry checking it in the formatter since that would be a simple check

File: test/phoenix_live_view/html_tokenizer_test.exs (hunk 2 above)
Comment (feliperenan) [reply test/phoenix_live_view/html_tokenizer_test.exs:109]: Yep. I will drop this test, it was just to exemplify that case you mentioned and help me to understand what would happen :)

File: lib/phoenix_live_view/html_tokenizer.ex [hunk 5]
```diff
@@ -94,15 +94,20 @@ defmodule Phoenix.LiveView.HTMLTokenizer do
...
 
   defp handle_text("<!--" <> rest, line, column, buffer, acc, state) do
+    state = update_in(state.context, &[:comment_start | &1])
     handle_comment(rest, line, column + 4, ["<!--" | buffer], acc, state)
   end
 
   defp handle_text("</" <> rest, line, column, buffer, acc, state) do
-    handle_tag_close(rest, line, column + 2, text_to_acc(buffer, acc, line, column), state)
+    text_to_acc = text_to_acc(buffer, acc, line, column, state.context)
+    state = put_in(state.context, [])
```
Comment (feliperenan) [inline lib/phoenix_live_view/html_tokenizer.ex:103]: We have to clean the state before putting the buffer to `acc` in order to avoid adding the `context` to tokens that are not related to `:comment`

**Comment by feliperenan** [comment]
@josevalim I believe now I got what I wanted 🤞🏻 . I have also updated my PR there in the Formatter to rely on these changes.

I'm not 100% sure if the code looks ok but the behavior now is correct. Please, take a look and feel free to apply or do any change you may like.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
